database/service_db.go

```diff
@@ -64,52 +87,119 @@ func (d *LevelDBServiceDB) All() ([]*service.Service, error) {
-		return nil, err
-	}
-
-	return services, nil
+	return services, iter.Error()
 }
 
 // Delete deletes service from database.
-func (d *LevelDBServiceDB) Delete(id string) error {
+func (d *LevelDBServiceDB) Delete(idOrAlias string) error {
+	id, alias, err := d.getIDAndAlias(idOrAlias)
+	if err != nil {
+		return err
+	}
+	// delete alias
+	if alias != "" {
+		if err := d.aliases.Delete([]byte(alias), nil); err != nil {
+			return err
+		}
+	}
+	// delete service
 	return d.db.Delete([]byte(id), nil)
 }
 
 // Get retrives service from database.
-func (d *LevelDBServiceDB) Get(id string) (*service.Service, error) {
+func (d *LevelDBServiceDB) Get(idOrAlias string) (*service.Service, error) {
+	id, _, err := d.getIDAndAlias(idOrAlias)
+	if err != nil {
+		return nil, err
+	}
 	b, err := d.db.Get([]byte(id), nil)
 	if err != nil {
-		if err == leveldb.ErrNotFound {
-			return nil, &ErrNotFound{ID: id}
-		}
 		return nil, err
 	}
-
-	return d.unmarshal(id, b)
+	return d.unmarshal(idOrAlias, b)
 }
 
 // Save stores service in database.
 func (d *LevelDBServiceDB) Save(s *service.Service) error {
+	// check service
 	if s.ID == "" {
-		return errors.New("database: can't save service without id")
+		return errCannotSaveWithoutID
 	}
+	// encode service
 	b, err := d.marshal(s)
 	if err != nil {
 		return err
 	}
-
-	return d.db.Put([]byte(s.ID), b, nil)
+	// save service
+	if err := d.db.Put([]byte(s.ID), b, nil); err != nil {
+		return err
+	}
+	// save alias if exist
+	if s.Alias != "" {
+		if err := d.aliases.Put([]byte(s.Alias), []byte(s.ID), nil); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
 // Close closes database.
 func (d *LevelDBServiceDB) Close() error {
-	return d.db.Close()
+	if err := d.db.Close(); err != nil {
+		return err
+	}
+	return d.aliases.Close()
+}
+
+// getIDAndAlias returns and separates id from alias.
+// id cannot be empty if an error is returned.
+// returned alias can be empty.
+func (d *LevelDBServiceDB) getIDAndAlias(idOrAlias string) (id string, alias string, err error) {
+	// check if idOrAlias is an id
+	// if it is already an id, return it with correspond alias if exist
+	isID, err := d.db.Has([]byte(idOrAlias), nil)
+	if err != nil {
+		return "", "", err
+	}
+	if isID {
+		id = idOrAlias
+		alias, err = d.getAliasFromID(id)
+		if err != nil {
+			return "", "", err
+		}
+		return id, alias, nil
+	}
+	// check if idOrAlias is an alias
+	// if it is an alias, return it with the corresponding id
+	alias = idOrAlias
+	idB, err := d.aliases.Get([]byte(alias), nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return "", "", &ErrNotFound{ID: alias}
+		}
+		return "", "", err
+	}
+	id = string(idB)
+	return id, alias, nil
+}
+
+// getAliasFromID returns the alias pointing to the id if exist.
+func (d *LevelDBServiceDB) getAliasFromID(id string) (alias string, err error) {
+	iter := d.aliases.NewIterator(nil, nil)
+	for iter.Next() {
+		if string(iter.Value()) == id {
+			alias = string(iter.Key())
```

Comment:
we can directly return here and release iteration with defer before starting to iterate